wordle-web/scripts/game.ts

```diff
@@ -1,74 +1,123 @@
-    console.log(this.secretWord);
-
-    // Populate guesses with the correct number of empty words
-    this.guesses = [];
-    for (let i = 0; i < this.maxAttempts; i++) {
-      this.guesses.push(
-        new Word({ maxNumberOfLetters: this.secretWord.length })
-      );
-    }
+    this.guessedLetters = [];
+    this.secretWord = WordList[Math.floor(Math.random() * WordList.length)].toUpperCase();
+    this.guesses = Array.from({ length: this.maxAttempts }, () => new Word({ maxNumberOfLetters: this.secretWord.length }));
+    console.log("Secret word: ", this.secretWord);
   }
 
-  public get guess() {
+  get currentGuess(): Word {
     return this.guesses[this.guessIndex];
   }
 
-  public removeLastLetter() {
+  public removeLastLetter(): void {
     if (this.gameState === GameState.Playing) {
-      this.guess.removeLastLetter();
+      this.currentGuess.removeLastLetter();
     }
   }
 
-  public addLetter(letter: string) {
-    if (this.gameState === GameState.Playing) {
-      this.guess.addLetter(letter);
+  public addLetter(letter: string): void {
+    if (this.gameState === GameState.Playing && letter.length === 1) {
+      this.currentGuess.addLetter(letter);
+      this.updateGuessedLetters();
     }
   }
 
-  public submitGuess() {
-    if (this.gameState !== GameState.Playing) return;
-    if (!this.guess.isFilled()) return;
-    if(!this.guess.isValidWord()){
-      this.guess.clear();
+  public submitGuess(): void {
+    if (this.gameState !== GameState.Playing || !this.currentGuess.isFilled || !this.currentGuess.isValidWord()) {
       return;
     }
 
-    if (this.guess.compare(this.secretWord)) {
+    const isCorrect = this.currentGuess.compare(this.secretWord);
+    this.updateGuessedLetters();
+
+    if (isCorrect) {
       this.gameState = GameState.Won;
+    } else if (this.guessIndex >= this.maxAttempts - 1) {
+      this.gameState = GameState.Lost;
     } else {
-      if (this.guessIndex === this.maxAttempts - 1) {
-        this.gameState = GameState.Lost;
-      } else {
-        this.guessIndex++;
-      }
+      this.guessIndex++;
     }
   }
+
+  public validWords(): string[] {
+    console.log("Recalculating valid words...");
+    return WordList.filter((word) => !this.isWordInvalid(word));
+  }
+
+  private isWordInvalid(word: string): boolean {
+    word = word.toLowerCase();
+    return this.guessedLetters.some((guessedLetter) => {
+      const letter = guessedLetter.char.toLowerCase();
+      const index = this.guesses[this.guessIndex].letters.findIndex(l => l.char.toLowerCase() === letter);
+      const isCorrect = guessedLetter.state === LetterState.Correct && word[index] !== letter;
+      const isMisplaced = guessedLetter.state === LetterState.Misplaced && (!word.includes(letter) || word[index] === letter);
+      const isWrong = guessedLetter.state === LetterState.Wrong && word.includes(letter);
+      return isCorrect || isMisplaced || isWrong;
```

Comment:
Just by reading this, it would seem that this would always be true, "is correct or is misplaced or is wrong"... however that doesn't appear to be the case, as the "spellbook" does indeed filter down.
1. Self-documenting-code is absolutely great, and your (possibly unintended) writing of it was as well, it's a valuable skill that is often overlooked, so good job.
2. However, this implementation was a bit off, and would have greatly benefited from a some unit tests, to dial in the accuracy.